### PR TITLE
Prefer node16 action to support macOS and Windows runners

### PR DESCRIPTION
**Pull request by umireon**
Implemented the same functionality as fetch_github_asset.sh in index.ts and made this action node16 action.
By this change, macOS and Windows runners on GitHub Actions will be supported.

### CODE REVIEW

**Comment by umireon** [comment]
Linux runner: https://github.com/umireon/fetch-gh-release-asset-test/runs/5672319585?check_suite_focus=true
macOS runner: https://github.com/umireon/fetch-gh-release-asset-test/runs/5672319629?check_suite_focus=true

**Comment by dsaltares** [comment]
Thanks for your contribution, it's really significant.

I hadn't given any thought to supporting Windows and MacOS runners. That sounds great for sure!

However, it does have significant implications as it changes the language of this repo from bash scripting to Typescript. I'm actually more familiar with TS than bash but the change needs to be reviewed a bit more closely as it's, essentially, a full rewrite. It may have different implications for other contributors.

Please bear with me while I go through it. Given personal circumstances, it may take me over a week.

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  "parser": "@typescript-eslint/parser",
+  "parserOptions": {
+    "ecmaVersion": "latest",
+    "sourceType": "module"
+  },
+  "plugins": [
+    "@typescript-eslint"
+  ],
+  "rules": {
+    "sort-keys": "error",
```
Comment (dsaltares) [inline .eslintrc.json:18]: Once this is merged, I may make some changes to the eslint config and potentially add prettier.

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+interface FetchAssetFileOptions {
+  readonly id: number
+  readonly outputPath: string
+  readonly owner: string
+  readonly repo: string
+}
+
+const MAX_RETRY = 5
+const RETRY_INTERVAL = 1000
```
Comment (dsaltares) [inline index.ts:53]: In the future we may want to replace the manual retry implementation with something like `async-retry`.

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+const main = async (): Promise<void> => {
+  const { owner, repo } = getRepo(core.getInput('repo', { required: false }), github.context)
+  const token = core.getInput('token', { required: false })
+  const version = core.getInput('version', { required: false })
+  const inputTarget = core.getInput('target', { required: false })
+  const file = core.getInput('file', { required: true })
+  const target = inputTarget === '' ? file : inputTarget
+
+  const octokit = github.getOctokit(token)
```
Comment (dsaltares) [inline index.ts:121]: This respects the defaulting behaviour here, right? https://github.com/dsaltares/fetch-gh-release-asset/pull/36/files#diff-fab4d7fb461bc6fbe9587f6c03fff98102b1c744145edcf2a993f2ff7cb05a0dR22.

**Review by dsaltares — COMMENTED** [review]
I just have 1 question really, but it looks pretty good.
I will do some refactoring later on and finally, add unit tests! Once it's merged.

**Comment by dsaltares** [comment]
@umireon It does look like the integration test is failing, would you be able to look into it please? Thanks!

**Comment by umireon** [comment]
@dsaltares It seems that the GitHub authorization mechanism is broken now. Please re-run the test later.

File: index.ts (hunk 3 above)
Comment (umireon) [reply index.ts:121]: I intended to implement the same behavior as the bash script.

File: index.ts (hunk 2 above)
Comment (umireon) [reply index.ts:53]: Great idea!

**Comment by umireon** [comment]
The existing Docker action is also broken now. The problem should be in the GitHub infrastructure.
https://github.com/umireon/fetch-gh-release-asset-test/runs/5737474231?check_suite_focus=true

**Comment by dsaltares** [comment]
@umireon GH's status page is green https://www.githubstatus.com and the error is related to API rate limiting. 

image

**Comment by umireon** [comment]
@dsaltares It hits the API limit due to the requests are not authorized against the GitHub API despite the correct Authorization header being present.
I'm pretty sure this is a GitHub outage because this code worked yesterday without errors many times.
I will ask for GitHub support for this phenomenon.

**Comment by umireon** [comment]
@dsaltares I've created a support ticket to GitHub. Please wait for the response from GitHub support.

**Comment by umireon** [comment]
Still no response from GitHub. I created the ticket from the paid organization so the response would be soon.

**Comment by umireon** [comment]
@dsaltares I talked with GitHub Support and managed to fix the problems.
Please re-run the integration tests.

**Comment by dsaltares** [comment]
Great stuff, thanks @umireon!

**Comment by guilhermeblanco** [comment]
This PR merge broke our pipeline.
It seems that since we now assume a well-formed version as part of the `getRelease` function, it completely ignored the scenario where you may be downloading draft assets (which means release version is "untagged_XXX").

What we experience now is a "Malformed version" error when attempting to download assets, as shown in the picture below.
Screenshot from 2022-04-19 10-31-15

**Comment by guilhermeblanco** [comment]
> @guilhermeblanco apologies, can you open an issue for this? In the meantime, can you pin your pipeline to the latest release instead of `master`? That will unblock you.

Done. https://github.com/dsaltares/fetch-gh-release-asset/issues/38

**Comment by umireon** [comment]
#40 will fix #38 sorry for inconvinience

**Pull request merged**